### PR TITLE
Refactored logging API to make use of libvlcLogInterop

### PR DESCRIPTION
**Pull request by jeremyVignelles**
This new API lets .net user receive logs from callbacks, thanks to https://github.com/jeremyVignelles/libvlcLogInterop .

Things left to do:

- [ ] Test libvlcLogInterop on other platforms.
  - [ ] Android
  - [ ] iOS
  - [ ] macOS
  - [ ] windows x86
- [ ] Publish packages for libvlcLogInterop

Tested on :

- [x] linux x64
- [x] windows x64

### CODE REVIEW

File: LibVLCSharp/Shared/Core.cs [hunk 1]
```diff
@@ -85,7 +150,46 @@ static IntPtr PreloadNativeLibrary(string nativeLibrariesPath, string libraryNam
...
+            {
+                return LinuxNative.dlopen(libraryPath, LinuxNative.RTLD_LAZY);
+            }
+            else
+            {
+                throw new PlatformNotSupportedException();
+            }
+        }
+
+        internal static T GetLibraryFunction<T>(IntPtr libraryHandle, string functionName)
```
Comment (mfkl) [inline LibVLCSharp/Shared/Core.cs]: I think these definitions belong to `MarshalUtils`

File: LibVLCSharp/Shared/Core.cs [hunk 2]
```diff
@@ -15,16 +15,56 @@ struct Native
...
+
+            public const int RTLD_LAZY = 1;
+            [DllImport("libdl", CallingConvention = CallingConvention.Cdecl)]
+            public static extern IntPtr dlopen(string lpFileName, int flags);
+
+            [DllImport("libdl", CharSet = CharSet.Ansi, CallingConvention = CallingConvention.Cdecl)]
+            public static extern string dlerror();
+        }
+
+        internal static IntPtr LibvlccoreHandle { get; private set; }
```
[the remark this answers is not in the document]
Comment (jeremyVignelles) [reply LibVLCSharp/Shared/Core.cs]: OK

File: LibVLCSharp/Shared/Core.cs (hunk 1 above)
Comment (jeremyVignelles) [reply LibVLCSharp/Shared/Core.cs]: then `PreloadNativeLibrary` too.

I think the loading architecture could be done in a separate class if you want, but maybe it's another PR? Maybe that other PR could be done before this one?
I've mixed a few things here, the linux .so loading and the real implementation of the logging API.
I'm just rebasing for now, will see in the future what to do with that.

File: LibVLCSharp/Shared/LibVLC.cs [hunk 3]
```diff
@@ -791,8 +771,11 @@ public enum LogLevel
     public delegate void ExitCallback();
 
     [SuppressUnmanagedCodeSecurity, UnmanagedFunctionPointer(CallingConvention.Cdecl)]
-    public delegate void LogCallback(IntPtr data, LogLevel logLevel, IntPtr logContext,
-        [MarshalAs(UnmanagedType.LPStr)] string format, IntPtr args);
+    internal delegate void LogCallback(IntPtr data, LibVLCLogLevel logLevel, IntPtr logContext,
+        IntPtr format, IntPtr args);
 
+    [SuppressUnmanagedCodeSecurity, UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+    internal delegate void FormattedLogCallback(IntPtr data, LibVLCLogLevel logLevel, IntPtr logContext,
```
[the remark this answers is not in the document]
Comment (jeremyVignelles) [reply LibVLCSharp/Shared/LibVLC.cs]: Yes, `FormattedLogCallback` will only ever correspond to LibVLC.OnLogInternal, which dispatches the log message on another thread.

End users will use `SetLogCallback` with their eventHandler

File: LibVLCSharp/Shared/LibVLC.cs [hunk 4]
```diff
@@ -377,6 +388,39 @@ public void SetAppId(string id, string version, string icon)
...
         }
 
+        public void SetLogCallback(Action<LogMessage> eventHandler, LibVLCLogLevel minLogLevel)
+        {
+            if(Core.LibvlcLogInteropHandle == IntPtr.Zero)
+            {
+                throw new InvalidOperationException("libvlcLogInterop library not loaded. Please refer to libvlcsharp's documentation to know how to install it.");
+            }
+
+            if(this._logCallbackReference != null)
```
[the remark this answers is not in the document]
Comment (jeremyVignelles) [reply LibVLCSharp/Shared/LibVLC.cs]: I'm used to having `this.` everywhere. I really don't care, but don't hesitate to tell me where I have this habits.

File: LibVLCSharp/Shared/Utf8StringMarshaler.cs [hunk 5]
```diff
@@ -63,5 +63,10 @@ public static ICustomMarshaler GetInstance()
         {
             return _instance;
         }
+
+        public static ICustomMarshaler GetInstance(string cookie)
```
[the remark this answers is not in the document]
Comment (jeremyVignelles) [reply LibVLCSharp/Shared/Utf8StringMarshaler.cs]: It's how ICustomMarshaler should be implemented. However, ICustomMarshaler is not available in netstandard1.1, and then it's useless, and we need to use the bulky shit of using manually the UTF8 marshaller.

**Comment by jeremyVignelles** [comment]
This seems to work as expected with #57 , closing this as no longer required.

**Pull request closed without merging**